src/mds/objects.py

```diff
@@ -391,43 +392,63 @@ async def delete_object(
+    fence DELETE /data/file_id by using an additional query parameter `delete_file_locations`.  
+    Uses the response status code from fence/indexd to determine whether user has
     permission to delete metadata.
 
     Args:
         guid (str): indexd GUID or alias, or MDS key
         request (Request): starlette request (which contains reference to FastAPI app)
     Returns:
         204: if record and metadata are deleted
-        403: if fence returns a 403 unauthorized response
-        500: if fence does not return 204 or 403 or there is an error deleting metadata
+        403: if fence/indexd returns a 403 unauthorized response
+        500: if fence/indexd does not return 204 or 403 or there is an error deleting metadata
     """
+    delete_file_locations = False
+    if "delete_file_locations" in request.query_params:
+        if request.query_params["delete_file_locations"]:
+            raise HTTPException(
+            HTTP_400_BAD_REQUEST, f"Query param `delete_file_locations` should not contain any value"
+        )
+        delete_file_locations = True
+    svc_name = "fence" if "delete_file_locations" in request.query_params else "indexd"
     try:
-        fence_endpoint = urljoin(config.DATA_ACCESS_SERVICE_ENDPOINT, f"data/{guid}")
         auth_header = str(request.headers.get("Authorization", ""))
         headers = {"Authorization": auth_header}
-        fence_response = await request.app.async_client.delete(
-            fence_endpoint, headers=headers
-        )
-        fence_response.raise_for_status()
+        if delete_file_locations:
+            fence_endpoint = urljoin(config.DATA_ACCESS_SERVICE_ENDPOINT, f"data/{guid}")
+            response = await request.app.async_client.delete(
+                fence_endpoint, headers=headers
+            )  
+        else:
+            rev = await get_indexd_revision(guid, request)
+            indexd_endpoint = urljoin(config.INDEXING_SERVICE_ENDPOINT, f"index/{guid}")
+            response = await request.app.async_client.delete(
+                indexd_endpoint, params = {"rev": rev}, headers=headers
+            )
+        response.raise_for_status()
         await (
-            Metadata.delete.where(Metadata.guid == guid)
-            .returning(*Metadata)
-            .gino.first()
+        Metadata.delete.where(Metadata.guid == guid)
+        .returning(*Metadata)
+        .gino.first()
         )
         return JSONResponse({}, HTTP_204_NO_CONTENT)
     except httpx.HTTPError as err:
         logger.debug(err)
         if err.response:
             raise HTTPException(
-                err.response.status_code, {"fence_response": err.response.text}
+                err.response.status_code, {f"{svc_name}_response": err.response.text}
             )
         raise HTTPException(
-            HTTP_500_INTERNAL_SERVER_ERROR, "Error during request to fence"
+            HTTP_500_INTERNAL_SERVER_ERROR, f"Error during request to {svc_name}"
         )
```

Comment:
for both exceptions here, we should log the actual exception, and imo return an error message that is _not_ the actual exception, so we don't leak internal information to users